### PR TITLE
[Messenger][AmqpExt] Add a retry mechanism for AMQP messages

### PR DESCRIPTION
**Pull request by sroze**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | ø
| License       | MIT
| Doc PR        | symfony/symfony-docs#9756

One of the most asked features (and required to have an enterprise-grade management of errors). I chose the approach of an AmqpExt-only retry mechanism to ensure we have a great user experience with this one at least. There are discussions on-going (#26945) to have more generic things like that across adapters. 

---

# So what's the point? 

When processing some messages, some error might happen. It might be that one of our 3rd party is down, that something is wrong with the last deployment, etc... All of these things will throw exceptions in the message handlers. The way of tackling this problem is very well described in @odolbeau's [Swarrot Retry provider real-life example](https://github.com/swarrot/swarrot/tree/master/src/Swarrot/Processor/Retry#real-example).

# How to configure it

Same than for the rest, this can be configured both via the DSN or the `options` when creating the adapter. Here is a bunch of working DSNs with their corresponding effect:

1. **Enable 3 retries** before having an exception
```yaml
amqp://guest:guest@localhost:5672/%2f/messages
    ?retry[attempts]=3
```

2. **Enable 3 retries** before message to be queue in a "dead queue" (i.e. messages to be looked at manually later)
```yaml
amqp://guest:guest@localhost:5672/%2f/messages
    ?retry[attempts]=3
    &retry[dead_queue]=dead_queue
```

3. **Enable 3 retries, dead queue and custom settings**
```yaml
amqp://guest:guest@localhost:5672/%2f/messages
    ?retry[attempts]=3
    &retry[exchange_name]=retry
    &retry[routing_key_pattern]=retry_attempt_%25attempt%25
    &retry[queue_name_pattern]=retry_queue_%attempt%
    &retry[dead_queue]=dead_queue
    &retry[dead_routing_key]=dead_routing_key
```

4. **Enable 3 retries, with custom TTLs per attempt.** Waits 10 seconds before retrying for the first attempt, 30 seconds for the 2nd attempt and 5 minutes for the 3rd one.
```yaml
amqp://guest:guest@localhost:5672/%2f/messages
    ?retry[attempts]=3
    &retry[ttl][0]=10000
    &retry[ttl][1]=30000
    &retry[ttl][2]=300000
```

### CODE REVIEW

File: src/Symfony/Component/Messenger/Adapter/AmqpExt/Connection.php [hunk 1]
```diff
@@ -101,6 +101,86 @@ public function publish(string $body, array $headers = array()): void
...
+        return $exchange;
+    }
+
+    private function retryQueue(string $name, int $attemptNumber)
+    {
+        $queue = $this->amqpFactory->createQueue($this->channel());
+        $queue->setName(str_replace('%attempt%', $attemptNumber, $name));
+        // $queue->setFlags(AMQP_DURABLE);
+        $queue->setArguments(array(
+            'x-message-ttl' => 30000, // 30 seconds
```
Comment (sroze) [inline src/Symfony/Component/Messenger/Adapter/AmqpExt/Connection.php]: This TTL should be configurable per attempt number.

File: src/Symfony/Component/Messenger/Adapter/AmqpExt/Connection.php [hunk 2]
```diff
@@ -101,6 +101,86 @@ public function publish(string $body, array $headers = array()): void
...
+            $queue->bind($exchange->getName(), 'dead');
+        }
+    }
+
+    private function retryExchange(array $retryConfiguration, int $attemptNumber)
+    {
+        $exchange = $this->amqpFactory->createExchange($this->channel());
+        $exchange->setName($retryConfiguration['name'] ?? 'retry');
+        $exchange->setType(AMQP_EX_TYPE_DIRECT);
+        // $exchange->setFlags(AMQP_DURABLE);
```
Comment (sroze) [inline src/Symfony/Component/Messenger/Adapter/AmqpExt/Connection.php]: To be removed.

File: src/Symfony/Component/Messenger/Adapter/AmqpExt/Connection.php [hunk 3]
```diff
@@ -101,6 +101,86 @@ public function publish(string $body, array $headers = array()): void
...
+
+        return $exchange;
+    }
+
+    private function retryQueue(string $name, int $attemptNumber)
+    {
+        $queue = $this->amqpFactory->createQueue($this->channel());
+        $queue->setName(str_replace('%attempt%', $attemptNumber, $name));
+        // $queue->setFlags(AMQP_DURABLE);
```
Comment (sroze) [inline src/Symfony/Component/Messenger/Adapter/AmqpExt/Connection.php]: To be removed.

File: src/Symfony/Component/Messenger/Adapter/AmqpExt/Connection.php [hunk 4]
```diff
@@ -101,6 +101,86 @@ public function publish(string $body, array $headers = array()): void
...
+
+        $maximumAttempts = $retryConfiguration['attempts'] ?? 3;
+        $routingKey = str_replace('%attempt%', $attemptNumber, $retryConfiguration['routing_key_pattern'] ?? 'attempt_%attempt%');
+
+        if ($attemptNumber > $maximumAttempts) {
+            if (!isset($retryConfiguration['dead_queue'])) {
+                return false;
+            }
+
+            $routingKey = 'dead';
```
Comment (sroze) [inline src/Symfony/Component/Messenger/Adapter/AmqpExt/Connection.php]: Should be a configurable `dead_routing_key`

File: src/Symfony/Component/Messenger/Adapter/AmqpExt/Connection.php [hunk 5]
```diff
@@ -101,6 +101,86 @@ public function publish(string $body, array $headers = array()): void
...
+
+        $queue = $this->retryQueue($retryConfiguration['queue_name_pattern'] ?? 'retry_queue_%attempt%', $attemptNumber);
+        $queue->declareQueue();
+        $queue->bind($exchange->getName(), str_replace('%attempt%', $attemptNumber, $retryConfiguration['routing_key_pattern'] ?? 'attempt_%attempt%'));
+
+        if (isset($retryConfiguration['dead_queue'])) {
+            $queue = $this->amqpFactory->createQueue($this->channel());
+            $queue->setName($retryConfiguration['dead_queue']);
+            $queue->declareQueue();
+            $queue->bind($exchange->getName(), 'dead');
```
Comment (sroze) [inline src/Symfony/Component/Messenger/Adapter/AmqpExt/Connection.php]: Should be using the `dead_routing_key` configuration.

File: src/Symfony/Component/Messenger/Adapter/AmqpExt/Connection.php (hunk 4 above)
Comment (sroze) [reply src/Symfony/Component/Messenger/Adapter/AmqpExt/Connection.php]: Well, that’s the term used by AMQP unfortunately. Maybe `too_many_attempts` if really you want to change? 🤔

File: src/Symfony/Component/Messenger/Adapter/AmqpExt/Connection.php [hunk 6]
```diff
@@ -101,6 +106,91 @@ public function publish(string $body, array $headers = array()): void
...
+
+        return $this->amqpRetryExchange;
+    }
+
+    private function retryQueue(array $retryConfiguration, int $attemptNumber)
+    {
+        $queue = $this->amqpFactory->createQueue($this->channel());
+        $queue->setName(str_replace('%attempt%', $attemptNumber, $retryConfiguration['queue_name_pattern'] ?? 'retry_queue_%attempt%'));
+        $queue->setArguments(array(
+            'x-message-ttl' => $retryConfiguration['ttls'][$attemptNumber - 1] ?? 30000, // 30 seconds by default
```
Comment (lyrixx) [inline src/Symfony/Component/Messenger/Adapter/AmqpExt/Connection.php]: why did you add a `s` in the key name ?

**Review by lyrixx — COMMENTED** [review]
IMHO, this really need to be tested against a real implementation of AMQP (RabbitMQ). 

I read very quickly the code and it seems this does not work well. I guess I should test it (by hand) to ensure work, but it looks like many case are not supported:

* retrying a message that come from an exchange in a fan out mode => the message will be dispatched is all queue that are bound. It should not
* retrying a message that come from an exchange in an header mode => the message will not be dispatched at all.

Then, I would be nice to be able to configure the type of retry strategy (linear / backoff), and offset (the 3/4 first iteration may be too close, and so we want to skip it) and of course the delay

Another things: all header should be casted to string.
From PHP, with the PECL, it's not really possible to create a typed header (in AMQP you can tell if the header is a string or an int). So it's better to use `string` everywhere to avoid edge case. I some version of rabbitmq, we hit this issue. The exchange were created with "string" type. But the message where created with "int" type. So there were not matches. Really hard to find bugs ;)

---

Edited: I forgot to say, you can grad some code from #23315 where the retry mechanisme is really advanced / configurable. But the more I read code about Messenger, The more I think we should reconsider #23315. How I see things: I should remove everything about worker, keep only AMQP stuff, and update the Messenger to use AMQP. (ping @fabpot)

**Comment by sroze** [comment]
> retrying a message that come from an exchange in a fan out mode => the message will be dispatched is all queue that are bound. It should not

That's the setup I used to test this retry mechanism and it works well. I'd love you to try and see really in which case it doesn't work  :)

> I think we should reconsider #23315. How I see things: I should remove everything about worker, keep only AMQP stuff, and update the Messenger to use AMQP.

If you believe you can't make the AMQP adapter much better by doing so, it is probably an interesting option (lots of work on your side though 😄). 

Though, there is a point in keeping the AMQP adapter simple and not having it "fully blown featured". Obviously, there is the maintenance cost point. But also, most people will use the default auto-setuped queues, right? I also believe that everybody willing to configure the queues and exchanges very precisely will very rarely be fully satisfied by auto-configuration (regardless of the amount of effort we put into it) and end up configuring their AMQP bits manually, that's what everybody has been doing so far at the end, and this retry mechanism PR allows you to configure the routing key pattern without auto-configuring it, so this use-case would work.

**Comment by kbond** [comment]
Question about retries in general: If a consumed message is handled by multiple handlers but only one throws an exception - how do you handle this? It probably isn't desirable to retry the message on handlers that were successful.

**Comment by makasim** [comment]
You could use [RabbitMqDlxDelayStrategy](https://github.com/php-enqueue/amqp-tools/blob/master/RabbitMqDlxDelayStrategy.php) as a source of inspiration for your retry version. 

It covers all scenarios and edge cases we could think of.

**Comment by nicolas-grekas** [comment]
Rebase needed after #27129

**Comment by sroze** [comment]
> If a consumed message is handled by multiple handlers but only one throws an exception - how do you handle this? It probably isn't desirable to retry the message on handlers that were successful.

That's a good question. You are right, it might not be desirable to retry on successful handlers but I believe that's the only (without a crazy amount of code) way to make it simple. On the other hand, I'd argue that something "critical" (a command in the CQRS point-of-view) will only have one handler. The things that will have multiple handlers are basically event subscribers and will be projections (or similar) and there should be no problem with having them running multiple times.

**Comment by sroze** [comment]
PR rebased.

**Comment by sroze** [comment]
I suggest we go ahead with this simple (optional) retry implementation for 4.1 and replace it with the one in the AMQP component later when introduced.

File: src/Symfony/Component/Messenger/Transport/AmqpExt/Connection.php [hunk 7]
```diff
@@ -101,6 +106,91 @@ public function publish(string $body, array $headers = array()): void
...
+            $queue->declareQueue();
+            $queue->bind($exchange->getName(), $retryConfiguration['dead_routing_key'] ?? 'dead');
+        }
+    }
+
+    private function retryExchange(array $retryConfiguration)
+    {
+        if (null === $this->amqpRetryExchange) {
+            $this->amqpRetryExchange = $this->amqpFactory->createExchange($this->channel());
+            $this->amqpRetryExchange->setName($retryConfiguration['name'] ?? 'retry');
```
Comment (ogizanagi) [inline src/Symfony/Component/Messenger/Transport/AmqpExt/Connection.php]: shouldn't it be `exchange_name` here?

File: src/Symfony/Component/Messenger/Transport/AmqpExt/Connection.php [hunk 8]
```diff
@@ -39,6 +39,11 @@ class Connection
      */
     private $amqpQueue;
 
+    /**
+     * @var \AMQPExchange|null
+     */
+    private $amqpRetryExchange;
+
     public function __construct(array $connectionCredentials, array $exchangeConfiguration, array $queueConfiguration, bool $debug = false, AmqpFactory $amqpFactory = null)
```
Comment (ogizanagi) [inline src/Symfony/Component/Messenger/Transport/AmqpExt/Connection.php]: IMHO all the config keys must be described in the docblock, as in https://github.com/symfony/symfony/blob/d264fcea24cc279054c7fe480e3dafb67f694668/src/Symfony/Component/Routing/Router.php#L105-L125 or https://github.com/symfony/symfony/blob/d264fcea24cc279054c7fe480e3dafb67f694668/src/Symfony/Component/HttpFoundation/Session/Storage/NativeSessionStorage.php#L63-L97 for instance

File: src/Symfony/Component/Messenger/Transport/AmqpExt/Connection.php [hunk 9]
```diff
@@ -101,6 +106,91 @@ public function publish(string $body, array $headers = array()): void
...
+
+        $retryConfiguration = $this->connectionCredentials['retry'];
+        $attemptNumber = ((int) $message->getHeader('symfony-messenger-attempts') ?: 0) + 1;
+
+        if ($this->shouldSetup()) {
+            $this->setupRetry($retryConfiguration, $attemptNumber);
+        }
+
+        $maximumAttempts = $retryConfiguration['attempts'] ?? 3;
+        $routingKey = str_replace('%attempt%', $attemptNumber, $retryConfiguration['routing_key_pattern'] ?? 'attempt_%attempt%');
```
Comment (ogizanagi) [inline src/Symfony/Component/Messenger/Transport/AmqpExt/Connection.php]: Could we normalize all the options values before using it, in the constructor for instance?

**Review by ogizanagi — DISMISSED** [review]
Not much to say. Looks good to me.

**Comment by kbond** [comment]
@sroze another potential problem could be an exception being thrown at a point where retrying the message would cause additional problems. 

What about an optional mechanism to have the subscriber tell the message bus/receiver to retry the message? This could also tell the bus to retry just this specific handler.

```php
// in your handler/subscriber
try {
    // ... some logic that throws an exception
} catch (\Throwable $e) {
    throw new RetryException(__CLASS__, $e); // or instead of __CLASS__ we get get the class from the exception's trace?
}
```

This is probably a discussion for another place as it is a more general retry - I can open an issue if we want to discuss further.

File: src/Symfony/Component/Messenger/Transport/AmqpExt/Connection.php (hunk 9 above)
Comment (sroze) [reply src/Symfony/Component/Messenger/Transport/AmqpExt/Connection.php]: Yep, it's a good idea.

**Comment by sroze** [comment]
I don't think that fabbot's PoV is relevant here :)

**Comment by sroze** [comment]
Documentation PR is here: symfony/symfony-docs/pull/9756

File: src/Symfony/Component/Messenger/Transport/AmqpExt/Connection.php [hunk 10]
```diff
@@ -101,6 +142,91 @@ public function publish(string $body, array $headers = array()): void
...
+
+        if (isset($retryConfiguration['dead_queue'])) {
+            $queue = $this->amqpFactory->createQueue($this->channel());
+            $queue->setName($retryConfiguration['dead_queue']);
+            $queue->declareQueue();
+            $queue->bind($exchange->getName(), $retryConfiguration['dead_routing_key']);
+        }
+    }
+
+    private function retryExchange(array $retryConfiguration)
```
Comment (Tobion) [inline src/Symfony/Component/Messenger/Transport/AmqpExt/Connection.php]: missing return type

File: src/Symfony/Component/Messenger/Transport/AmqpExt/AmqpReceiver.php [hunk 11]
```diff
@@ -62,9 +63,14 @@ public function receive(callable $handler): void
 
                 throw $e;
             } catch (\Throwable $e) {
-                $this->connection->nack($AMQPEnvelope, AMQP_REQUEUE);
+                if (!$this->connection->publishForRetry($AMQPEnvelope)) {
```
Comment (Tobion) [inline src/Symfony/Component/Messenger/Transport/AmqpExt/AmqpReceiver.php]: I think you need to handle errors here as well. Otherwise it kills the whole receiver.

File: src/Symfony/Component/Messenger/Transport/AmqpExt/Connection.php [hunk 12]
```diff
@@ -39,9 +39,50 @@ class Connection
...
+     *     * name: Name of the exchange
+     *     * type: Type of exchange (Default: fanout)
+     *     * flags: Exchange flags (Default: AMQP_DURABLE)
+     *     * arguments: Extra arguments
+     *   * retry:
+     *     * attempts: Number of times it will try to retry
+     *     * routing_key_pattern: The pattern of the routing key (Default: "attempt_%attempt%")
+     *     * dead_queue: Name of the queue in which messages that retry more than attempts time are pushed to
+     *     * dead_routing_key: Routing key name for the dead queue (Default: "dead")
+     *     * queue_name_pattern: Pattern to use to create the queues (Default: "retry_queue_%attempt%")
```
Comment (Tobion) [inline src/Symfony/Component/Messenger/Transport/AmqpExt/Connection.php]: The only point of using different queues with each attempt is to be able to use different ttls, isn't it? But if the ttl is the same (the default), there is no point in having diferent queues. So to me, the retry queues should be based on the ttl, not the attempt number.

File: src/Symfony/Component/Messenger/Transport/AmqpExt/Connection.php [hunk 13]
```diff
@@ -101,6 +142,91 @@ public function publish(string $body, array $headers = array()): void
...
+            }
+
+            $routingKey = $retryConfiguration['dead_routing_key'];
+        }
+
+        $this->retryExchange($retryConfiguration)->publish(
+            $message->getBody(),
+            $routingKey,
+            AMQP_NOPARAM,
+            array('headers' => array_merge($message->getHeaders(), array('symfony-messenger-attempts' => (string) $attemptNumber)))
```
Comment (Tobion) [inline src/Symfony/Component/Messenger/Transport/AmqpExt/Connection.php]: a message has more attributes than just the headers like delivery_mode. do those attributes not get lost if you don't forward them?

File: src/Symfony/Component/Messenger/Transport/AmqpExt/Connection.php (hunk 12 above)
Comment (sroze) [reply src/Symfony/Component/Messenger/Transport/AmqpExt/Connection.php]: That's correct, the different queues are used for different TTLs, but not only. On the operational side of the things (for whoever will monitor the RabbitMq), it's much clearer to understand what is going on in terms of retry, while it would be trickier with a `retry_in_X` IMHO)

File: src/Symfony/Component/Messenger/Transport/AmqpExt/Connection.php [hunk 14]
```diff
@@ -39,9 +39,50 @@ class Connection
...
     private $amqpQueue;
 
-    public function __construct(array $connectionCredentials, array $exchangeConfiguration, array $queueConfiguration, bool $debug = false, AmqpFactory $amqpFactory = null)
+    /**
+     * @var \AMQPExchange|null
+     */
+    private $amqpRetryExchange;
+
+    /**
+     * Available options:
```
[the remark this answers is not in the document]
Comment (ogizanagi) [reply src/Symfony/Component/Messenger/Transport/AmqpExt/Connection.php]: Yes, I think it's fine like this (but not ideal, true). Otherwise we'll need something as similar as the `SecurityFactoryInterface` in order to make it extendable to any transport with similar use-cases.

File: src/Symfony/Component/Messenger/Transport/AmqpExt/AmqpReceiver.php (hunk 11 above)
Comment (sroze) [reply src/Symfony/Component/Messenger/Transport/AmqpExt/AmqpReceiver.php]: Added

File: src/Symfony/Component/Messenger/Transport/AmqpExt/Connection.php (hunk 13 above)
Comment (sroze) [reply src/Symfony/Component/Messenger/Transport/AmqpExt/Connection.php]: Good point, added the ones documented by the AMQP extension.

**Comment by sroze** [comment]
ping @symfony/deciders. We should go ahead with this one in 4.1.

**Comment by weaverryan** [comment]
If this is a simple implementation, and we’ll use a “better” one later, should we merge it? What risks are there with this implementation? Is it maybe just not as robust as we would like? Would it be reasonable to deprecate later? And can you summarize the ongoing concerns from people.

Thanks for working on this - there have been quite a few things mentioned in comments, and I’m not sure what’s still relevant.

**Comment by sroze** [comment]
@javiereguiluz I moved the "hidden defaults" to constants.

**Comment by sroze** [comment]
That isn't a real success; let's move it to 4.2 then.

File: src/Symfony/Component/Messenger/Transport/AmqpExt/Connection.php [hunk 15]
```diff
@@ -204,9 +350,9 @@ public function exchange(): \AMQPExchange
         return $this->amqpExchange;
     }
 
-    public function getConnectionCredentials(): array
+    public function getConnectionConfiguration(): array
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Messenger/Transport/AmqpExt/Connection.php]: That's a BC break, but the component is experimental.
Still, should be mentioned in the upgrade/changelog files I suppose?

Comment (sroze) [reply src/Symfony/Component/Messenger/Transport/AmqpExt/Connection.php]: Yes, we can definitely mention this in the upgrade file!

Comment (sroze) [reply src/Symfony/Component/Messenger/Transport/AmqpExt/Connection.php]: @nicolas-grekas updated the CHANGELOG 👍

**Comment by sroze** [comment]
This pull-request is welcoming another round of review 💚

**Comment by Koc** [comment]
@sroze how can we programatically requeue message with specified delay from handler? For example for cases, when we cann't process this message now.

**Comment by sroze** [comment]
> can we programatically requeue message with specified delay from handler?

In the given implementation we can't. I believe that this is out of scope for now anyway and that it will be handled with the AMQP component (#27140) which will provide more features such as this one 👍

**Comment by fabpot** [comment]
@sroze Can we talk about this one this week IRL?

**Comment by ragboyjr** [comment]
@sroze do you think it would be possible to implement this in away so that it would standardize the retrying logic across the various transports?

**Comment by weaverryan** [comment]
I've brought the AMQP retry logic from Sam and put it into #30557  - but also fitting it into a system where retrying has been made generic, so it can be implemented/included in future transports.

I would LOVE review on that - I'm running full speed to get the features Messenger needs/deserves, but I need queue experts to help me avoid pitfalls.

**Comment by lyrixx** [comment]
IMHO this one could be closed in favor or https://github.com/symfony/symfony/pull/30557

**Comment by fabpot** [comment]
Closing in favor of #30557
